src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/utilities/HeapHprofBinWriter.java

```diff
@@ -472,47 +470,213 @@ public synchronized void write(String fileName) throws IOException {
 
         // flush buffer stream.
         out.flush();
-
         if (!useSegmentedHeapDump) {
-            // Fill in final length.
             fillInHeapRecordLength();
         } else {
-            hprofBufferedOut.finish();
             // Write heap segment-end record
             out.writeByte((byte) HPROF_HEAP_DUMP_END);
             out.writeInt(0);
             out.writeInt(0);
         }
-
         // flush buffer stream and throw it.
         out.flush();
         out.close();
         out = null;
         hprofBufferedOut = null;
+        currentSegmentStart = 0;
     }
 
-    @Override
-    protected void writeHeapRecordPrologue() throws IOException {
-        if (useSegmentedHeapDump) {
-            hprofBufferedOut.enterSegmentMode();
-        } else if (currentSegmentStart == 0) {
-            // write heap data header
-            out.writeByte((byte) (HPROF_HEAP_DUMP));
-            out.writeInt(0);
+    protected int calculateOopDumpRecordSize(Oop oop) throws IOException {
+        if (oop instanceof TypeArray) {
+            return calculatePrimitiveArrayDumpRecordSize((TypeArray)oop);
+        } else if (oop instanceof ObjArray) {
+            Klass klass = oop.getKlass();
+            ObjArrayKlass oak = (ObjArrayKlass) klass;
+            Klass bottomType = oak.getBottomKlass();
+            if (bottomType instanceof InstanceKlass ||
+                    bottomType instanceof TypeArrayKlass) {
+                return calculateObjectArrayDumpRecordSize((ObjArray)oop);
+            } else {
+                // internal object, nothing to write.
+                return 0;
+            }
+        } else if (oop instanceof Instance) {
+            Instance instance = (Instance) oop;
+            Klass klass = instance.getKlass();
+            Symbol name = klass.getName();
+            if (name.equals(javaLangString)) {
+                return calculateStringDumpRecordSize(instance);
+            } else if (name.equals(javaLangClass)) {
+                return calculateClassInstanceDumpRecordSize(instance);
+            } else if (name.equals(javaLangThread)) {
+                return calculateThreadDumpRecordSize(instance);
+            } else {
+                klass = klass.getSuper();
+                while (klass != null) {
+                    name = klass.getName();
+                    if (name.equals(javaLangThread)) {
+                        return calculateThreadDumpRecordSize(instance);
+                    }
+                    klass = klass.getSuper();
+                }
+                return calculateInstanceDumpRecordSize(instance);
+            }
```

Comment:
The code above is really confusing.
You have these methods defined (Q: why are these methods not defined close to each other?):
```
+    private int calculateStringDumpRecordSize(Instance instance) {
+        return calculateInstanceDumpRecordSize(instance);
+    }
. . .
+    private int calculateThreadDumpRecordSize(Instance instance) {
+        return calculateInstanceDumpRecordSize(instance);
+    }
```
So, the methods `calculateStringDumpRecordSize()` and `calculateThreadDumpRecordSize()` are the same as the `calculateInstanceDumpRecordSize()`.

Then you have the code:
```
+        } else if (oop instanceof Instance) {
+            Instance instance = (Instance) oop;
+            Klass klass = instance.getKlass();
+            Symbol name = klass.getName();
+            if (name.equals(javaLangString)) {
+                return calculateStringDumpRecordSize(instance);
+            } else if (name.equals(javaLangClass)) {
+                return calculateClassInstanceDumpRecordSize(instance);
+            } else if (name.equals(javaLangThread)) {
+                return calculateThreadDumpRecordSize(instance);
+            } else {
+                klass = klass.getSuper();
+                while (klass != null) {
+                    name = klass.getName();
+                    if (name.equals(javaLangThread)) {
+                        return calculateThreadDumpRecordSize(instance);
+                    }
+                    klass = klass.getSuper();
+                }
+                return calculateInstanceDumpRecordSize(instance);
+            }
+        } else {
```
You can simplify the above with something like this:
```
+        } else if (oop instanceof Instance) {
+            Instance instance = (Instance) oop;
+            Klass klass = instance.getKlass();
+            Symbol name = klass.getName();
+            if (name.equals(javaLangString) ||
+                name.equals(javaLangClass) ||
+                name.equals(javaLangThread)) {
+                return calculateClassInstanceDumpRecordSize(instance);
+            } else {
+                return calculateInstanceDumpRecordSize(instance);
+            }
+        } else {
```
And my guess is that this simplified code can be simplified even further (Q: is it right?): 
```
+        } else if (oop instanceof Instance) {
+            Instance instance = (Instance) oop;
+                return calculateInstanceDumpRecordSize(instance);
+            }
+        } else {
```
And the methods `calculateStringDumpRecordSize()` and `calculateThreadDumpRecordSize()` can be removed as unneeded.